src/main/java/bio/terra/workspace/service/resource/controlled/flight/CreateGcsBucketStep.java

```diff
@@ -1,28 +1,136 @@
     this.crlService = crlService;
+    this.resource = resource;
+    this.userRequest = userRequest;
   }
 
   @Override
   public StepResult doStep(FlightContext flightContext)
       throws InterruptedException, RetryException {
+
+    final BucketInfo bucketInfo =
+        BucketInfo.newBuilder(resource.getBucketName())
+            .setLocation(resource.getLocation())
+            .setStorageClass(ApiConversions.toGcsApi(resource.getDefaultStorageClass()))
+            .setLifecycleRules(ApiConversions.toGcsApi(resource.getLifecycle()))
+            .build();
+
+    final StorageCow storageCow = crlService.createStorageCow(userRequest);
+    final BucketCow bucketCow = storageCow.create(bucketInfo);
+
     return StepResult.getStepResultSuccess();
   }
 
   @Override
   public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    final StorageCow storageCow = crlService.createStorageCow(userRequest);
+    final boolean deleted = storageCow.delete(resource.getBucketName());
+
     return StepResult.getStepResultSuccess();
   }
+
+  private static class ApiConversions {
+
+    private ApiConversions() {}
+
+    private static StorageClass toGcsApi(GoogleBucketDefaultStorageClass storageClass) {
+      switch (storageClass) {
+        case STANDARD:
+          return StorageClass.STANDARD;
+        case NEARLINE:
+          return StorageClass.NEARLINE;
+        case COLDLINE:
+          return StorageClass.COLDLINE;
+        case ARCHIVE:
+          return StorageClass.ARCHIVE;
+        default:
+          throw new IllegalStateException("Unrecognized storage class " + storageClass);
+      }
+    }
+
+    private static List<LifecycleRule> toGcsApi(GoogleBucketLifecycle lifecycle) {
+      return lifecycle.getRules().stream()
+          .map(ApiConversions::toGcsApi)
+          .collect(Collectors.toList());
+    }
+
+    private static LifecycleRule toGcsApi(GoogleBucketLifecycleRule lifecycleRule) {
+      return new LifecycleRule(
+          toGcsApi(lifecycleRule.getAction()), toGcsApi(lifecycleRule.getCondition()));
+    }
+
+    private static LifecycleAction toGcsApi(GoogleBucketLifecycleRuleAction lifecycleRuleAction) {
+      switch (lifecycleRuleAction.getType()) {
+        case DELETE:
+          return LifecycleAction.newDeleteAction();
+        case SET_STORAGE_CLASS:
+          return LifecycleAction.newSetStorageClassAction(
+              toGcsApi(lifecycleRuleAction.getStorageClass()));
+        default:
+          throw new IllegalStateException(
+              "Unrecognized lifecycle action type " + lifecycleRuleAction.getType());
+      }
+    }
+
+    private static LifecycleCondition toGcsApi(GoogleBucketLifecycleRuleCondition condition) {
+      final LifecycleCondition.Builder resultBuilder = LifecycleCondition.newBuilder();
+
+      // TODO: some conditions aren't on the Google api object
+      Optional.ofNullable(condition.getAge()).ifPresent(resultBuilder::setAge);
```

Comment:
Reverse polish assignment! OMG!!

I don't think you have to do so much work.
```
resultBuilder.setAge(condition.getAge())
```
works just fine.

Reply:
It's natural once you get used to it. The alternative won't work if `setAge()` assumes a non-null argument. Probably not an issue in this case, assuming no side effects happen when setting this value.

Reply:
I just remembered the situation where this was necessary. For a BigQuery insert object, I had to make a map of all the non-null columns, as the map didn't support null. So this pattern was a good fit for that. I checked and this builder doesn't care if it gets nulls.